# 分组过滤，运算，排序

## 分组过滤

In [11]:
import numpy as np
import pandas as pd

In [12]:
a=pd.DataFrame({"a":["i","j","k","i","j","k"],"b":[2,1,4,1,2,4],"c":[3,2,2,1,2,3]})

In [13]:
a

a  b  c
0  i  2  3
1  j  1  2
2  k  4  2
3  i  1  1
4  j  2  2
5  k  4  3

In [14]:
for i in a.groupby("a"):
    print(i[1])
    print()

   a  b  c
0  i  2  3
3  i  1  1

   a  b  c
1  j  1  2
4  j  2  2

   a  b  c
2  k  4  2
5  k  4  3



In [5]:
#这里的x为已经groupby之后的数据
def gl(x):
    return x["b"].sum()>3

In [38]:
a.groupby("a").filter(gl)
#最终结果是满足条件的原始记录（注意结果的形式）
#请特别注意各个函数方法的左右效果

a  b  c
2  k  4  2
5  k  4  3

In [6]:
p=a.groupby("a").b.sum()
p[p>3]

a
k    8
Name: b, dtype: int64

## 课间题

* 请过滤出所有c列中值为偶数的记录(这里不用filter，直接用以前的方法过滤就行)

a  b  c
1  j  1  2
2  k  4  2
4  j  2  2

* 请过滤出c分组内记录行数大于等于2的数据

In [22]:
a

a  b  c
0  i  2  3
1  j  1  2
2  k  4  2
3  i  1  1
4  j  2  2
5  k  4  3

In [29]:
a.count().b

6

## 排序

* groupby通常会自动对行标排序

In [36]:
a1=pd.Series([2,3,6,1],index=["i","u","2","1"])#排序需要在同种类型的数据中进行（比如整数与字符就不能比较）
a1

i    2
u    3
2    6
1    1
dtype: int64

In [40]:
a1.sort_values(ascending=False)[0:2]

2    6
u    3
dtype: int64

In [42]:
a1.sort_index(ascending=False)

u    3
i    2
2    6
1    1
dtype: int64

In [43]:
b1=pd.DataFrame(np.random.randn(4,3),index=[2,4,1,3], columns=["b","a","k"])
b1

b         a         k
2 -0.571647 -1.125259  0.669289
4  0.289666 -0.642328  1.604716
1  0.517962  1.002307  0.217916
3  0.456894 -0.360221 -0.414614

In [44]:
b1.sort_index()

b         a         k
1  0.517962  1.002307  0.217916
2 -0.571647 -1.125259  0.669289
3  0.456894 -0.360221 -0.414614
4  0.289666 -0.642328  1.604716

In [12]:
b1.sort_index(axis=1)

a         b         k
2  0.749984  0.954464  0.668990
4  0.244264 -0.811406  0.234548
1  0.502138 -1.074694 -0.846381
3 -1.616720  1.394678 -0.313095

In [45]:
b1.sort_values(by="b")

b         a         k
2 -0.571647 -1.125259  0.669289
4  0.289666 -0.642328  1.604716
3  0.456894 -0.360221 -0.414614
1  0.517962  1.002307  0.217916

## 课间题

In [46]:
r1=pd.read_csv(r"D:\t_alibaba_data3.txt",names=["user","brand","behavr","date"],sep="\t",dtype={"behavr":int})
r1.head()

user  brand  behavr   date
0  10944750  13451       0  06/04
1  10944750  13451       2  06/04
2  10944750  13451       2  06/04
3  10944750  13451       0  06/04
4  10944750  13451       0  06/04

* 求各个商品的浏览总量

brand
11     4
15     6
18     1
19     2
20     5
22    33
23     1
26     9
27     4
29    21
dtype: int64

brand
11    4
15    6
18    1
19    2
20    5
dtype: int64

* 输出商品id最小的前10个商品的及其浏览量

brand
11     4
15     6
18     1
19     2
20     5
22    33
23     1
26     9
27     4
29    21
dtype: int64

* 输出商品浏览最大的前10个商品及其浏览量

brand
7868     3196
11196    2076
11080    1652
12220    1650
21110    1586
2257     1387
26722    1311
3228     1242
14261    1240
14020    1139
dtype: int64

* 输出用户10944750日期最后的20条浏览记录

## series与dataframe的运算

* 数据框和序列的运算本质上是一种自动遍历，用来避免直接python循环的好方法

* 运算函数：sub;add;div

* 与常数的运算

In [57]:
a=pd.Series([2,1,4,5],index=["a","b","c","d"])
a

a    2
b    1
c    4
d    5
dtype: int64

In [15]:
#和数值进行运算完全遵守numpy类似的规则：就是逐个作用到每一个元素
a+3

a    5
b    4
c    7
d    8
dtype: int64

In [16]:
np.exp(a)#无论是内部函数

a      7.389056
b      2.718282
c     54.598150
d    148.413159
dtype: float64

In [60]:
a**2+6

a    10
b     7
c    22
d    31
dtype: int64

In [56]:
def w1(x):#还是自定义函数
    return x-9+2

print(w1(a))

a   -5
b   -6
c   -3
d   -2
dtype: int64


In [61]:
d=pd.DataFrame(np.random.randn(3,2),columns=["a","b"])
d

a         b
0 -0.265667 -1.037581
1 -1.033648  0.443449
2  0.619189  1.741607

In [62]:
d-3

a         b
0 -3.265667 -4.037581
1 -4.033648 -2.556551
2 -2.380811 -1.258393

In [63]:
d.sub(3)

a         b
0 -3.265667 -4.037581
1 -4.033648 -2.556551
2 -2.380811 -1.258393

In [21]:
np.exp(d)

a         b
0  1.405256  3.363054
1  2.375396  1.037718
2  4.253211  0.323050

In [22]:
print(w1(d))

          a         b
0 -6.659781 -5.787151
1 -6.134836 -6.962976
2 -5.552326 -8.129949


## 课间题

* 求每天的总记录数

* 求平均日记录数，并给出每天相对于平均记录数的差值,然后对差值求绝对值(求绝对值函数为abs())

In [67]:
abs(-2.4)

2.4

1486.8292682926829

date
04/15     207.170732
04/16      88.829268
04/17     436.829268
04/18    1041.170732
04/19       3.170732
04/20     253.829268
04/21       4.170732
04/22     399.829268
04/23    1445.829268
04/24     832.829268
04/25     269.829268
04/26      50.829268
04/27     383.829268
04/28     243.829268
04/29     342.829268
04/30     715.829268
05/01     726.829268
05/02     452.829268
05/03     121.829268
05/04     137.829268
05/05     197.829268
05/06     222.829268
05/07     459.829268
05/08     496.829268
05/09      50.829268
05/10       7.170732
05/11     331.170732
05/12    1422.829268
05/13      65.829268
05/14     439.829268
            ...     
07/17     267.829268
07/18     133.170732
07/19     238.170732
07/20      87.170732
07/21      26.170732
07/22      90.829268
07/23     231.829268
07/24     271.829268
07/25     148.170732
07/26     602.170732
07/27     649.170732
07/28     785.170732
07/29     133.170732
07/30     196.170732
07/31      68.170732
08/01     206.170732
08/02   

## 与序列以及数据框的运算（索引对齐）

In [73]:
a=pd.Series([2,1,4,5],index=["a","b","c","d"])
a

a    2
b    1
c    4
d    5
dtype: int64

In [74]:
a+a

a     4
b     2
c     8
d    10
dtype: int64

In [75]:
b=pd.Series([3,1,4,2,1],index=["a","b","c","j","p"])
b

a    3
b    1
c    4
j    2
p    1
dtype: int64

In [76]:
#索引会取并集，找不到相同索引匹配进行运算的则为nan
a+b

a    5.0
b    2.0
c    8.0
d    NaN
j    NaN
p    NaN
dtype: float64

In [31]:
#如果你希望缺失值用其他值来替代而不是空缺，则需要用函数来运算并加入fill_value参数
a.add(b,fill_value=0)
#上面没找到匹配值的时候则用0来替代
#注意加法运算之后数据类型都变为了float

a    5.0
b    2.0
c    8.0
d    5.0
j    2.0
p    1.0
dtype: float64

In [77]:
d=pd.DataFrame(np.random.randint(0,10,size=(3,2)),columns=["a","b"])
d

a  b
0  3  3
1  8  4
2  0  2

In [78]:
c=pd.DataFrame(np.random.randint(0,10,size=(4,3)),columns=["a","b","c"])
c

a  b  c
0  5  7  8
1  8  4  8
2  8  3  5
3  3  2  2

In [82]:
d.div(c,fill_value=2)

a         b     c
0  0.600000  0.428571  0.25
1  1.000000  1.000000  0.25
2  0.000000  0.666667  0.40
3  0.666667  1.000000  1.00

## 课间题

* 求各个商品的购买量

* 求各个商品的浏览量

* 求各个商品的转化率（总购买比上总浏览）

brand
11       0.000000
15       0.000000
18       0.000000
19       0.500000
20       0.000000
22       0.030303
23       0.000000
26       0.000000
27       0.000000
29       0.142857
33       0.000000
38       0.000000
39       0.000000
43       0.125000
45       0.000000
53       0.000000
57       0.111111
58       0.000000
62       0.083333
72       0.000000
83       1.250000
90       0.000000
91       0.000000
103      0.000000
104      0.000000
111      0.000000
116      0.000000
119      0.000000
120      0.000000
121      0.000000
           ...   
29463    0.000000
29464    0.000000
29465    0.063492
29472    0.105263
29473    0.020408
29474    0.000000
29475    0.000000
29477    0.000000
29486    0.000000
29487    0.000000
29488    0.000000
29490    0.028169
29491    0.000000
29493    0.000000
29497    0.000000
29501    0.000000
29502    0.000000
29504    0.000000
29505    0.142857
29509    0.000000
29524    0.000000
29528    0.000000
29529    0.000000
29531    0.000000
2953

* 转化率最高的30个商品及其转化率

brand
16602         inf
27653         inf
28617         inf
526           inf
12018         inf
25349         inf
24819         inf
15959         inf
2291          inf
24434         inf
13017         inf
18124         inf
23342         inf
3051          inf
22306         inf
14174    5.000000
2208     5.000000
10690    3.000000
948      2.000000
17977    2.000000
6757     2.000000
25333    1.500000
2437     1.333333
83       1.250000
26168    1.222222
10308    1.200000
15828    1.000000
18542    1.000000
5478     1.000000
1704     1.000000
dtype: float64

* 可否同时输出这30个商品转化率，购买量，浏览量等信息（需要迭代）

16602 inf 2 0
27653 inf 1 0
28617 inf 1 0
526 inf 1 0
12018 inf 4 0
25349 inf 1 0
24819 inf 1 0
15959 inf 1 0
2291 inf 1 0
24434 inf 1 0
13017 inf 1 0
18124 inf 1 0
23342 inf 1 0
3051 inf 1 0
22306 inf 1 0
14174 5.0 5 1
2208 5.0 5 1
10690 3.0 3 1
948 2.0 2 1
17977 2.0 2 1
6757 2.0 2 1
25333 1.5 12 8
2437 1.3333333333333333 4 3
83 1.25 10 8
26168 1.2222222222222223 11 9
10308 1.2 6 5
15828 1.0 2 2
18542 1.0 1 1
5478 1.0 1 1
1704 1.0 1 1


* 求所有购买量大于20的各个商品的转化率(请自己百度去掉nan值的方法)

brand
554      0.117647
905      0.579710
911      0.043189
2257     0.017304
2367     0.105991
2683     0.077758
3228     0.030596
3627     0.035573
4949     0.022284
7105     0.032061
7868     0.038798
8637     0.079027
8689     0.043434
11080    0.016949
11196    0.032755
12168    0.044481
14020    0.036874
14261    0.032258
15019    0.090164
15584    0.053476
18024    0.093264
18805    0.039792
23779    0.111111
26288    0.095890
26722    0.020595
27117    0.283784
27791    0.073325
29099    0.050955
dtype: float64